Add ability to query pumphistory from a specific record id.

alimhassam opened the pull request:
Querying by timestamp is sometimes racy because some records have the
same timestamp such as Bolus and BolusWizard.
Also the timestamps are not always monotonically increasing when the
clock has been changed.

Being able to query the pumphistory from a specific record id avoids
these race conditions.

ecc1 commented:
Can you describe the problem this solves in more detail? In particular, why doesn't fetching records strictly more recent than a given time work? I'm not wild about including the whole (base64-encoded) record in each object uploaded to NS.

alimhassam commented:
I see, I don't think these would get uploaded to NS but I'll double check or add the appropriate filtering in oref0 NS upload (wherever that's done) if needed. In currently working on and testing the oref0 corresponding change.

The problem this solves is that it makes the incremental pumphistory updates much more robust and avoids many different race conditions with time bases queries.
The goal is to make sure that the incremental pumphistory update produces a pumphistory file with is identical than if we queried a full pump history.

I have 3 known examples of issues with time based queries that wouldn't be an issue with queries by record identifier.

Example 1: records with same timestamps. For example Bolus and Boluswizard.

The bolus and boluswizard records have the same timestamps (call that time TS). 
Say at time t0 we query the pumphistory and get only the boluswizard record but for some reason we don't get the corresponding bolus record.
At time t1 we do an increment pumphistory since TS the missing bolus record would be excluded since time TS is not strictly after TS.

Example 2: out of order records (by timestamp) due to time change.
Bolus 1 At 7:00am
Pumphistory. Latest viewed time is 7:00am
Bolus 2 at 7:01am
Change clock to 6:00am
Bolus 3 at 6:01am
Pumphistory since 7:00am incorrectly excludes bolus 2and3

Example 3: out of order records (by timestamp) due to other reasons:
DailyTotal records are not in increasing timestamp orders. Querying across day change used to stop processing records before and ignore records slightly before midnight. This is fixed with a different  approach but there is still
the risk of duplication of the entire day of records if the top records on pumphistory is the daily total record.


All these are solved by going with a unique record identifier making incremental pumphistory updates much more robust and guaranteed to reflect what's actually on the pump in the order the pump is storing the records.

Let me know if these examples are clear enough or if there is something I'm not seeing.

alimhassam commented:
Note there were reports of issues recently, possibly due to example 1 here on the ITB forum. That's what prompted me to work on this now.

ecc1 commented:
Thanks for the clear explanation, very helpful.

I did some quick experiments with CRCs over the record data. A CRC-16 resulted in a few collisions (across all of my test data), while a CRC-32 had no collisions. So an 8-byte string (hex version of the CRC-32) would almost certainly be enough as a record ID

alimhassam commented:
I'm not convinced the extra complexiting of adding a CRC is really needed, or what benefits it really gives. I also haven't looked up the risk of collisions etc, or how ofter we get long base64 id's.

We should still note that the effect of a collision might be incorrect pumphistory which we should really try to avoid (and is also the issue we're trying to solve here).

That said, if you think it's needed for some reasons (ie ids are too long, and the extra data is causing problems), and we can guarantee the risk of collision is so small that it's never going to be a problem then we could go for that approach.

We could minimise issues further by only applying the CRC when the ids are too long say above 32 characters or something un-frequent if you want.

alimhassam commented:
I have a follow up improvement i'd like to add, but i'm not sure exactly how to best do it. I'd like to  have a way to distinguish if pumphistory failed to find the `from` recordId, vs an issue with pump communication. 

In that first case, i would like oref0 to fallback to a full pump history refresh, whereas in that secord case, we should keep the existing behavior or just waiting for better pump communication.

ecc1 commented:
I was thinking it would be cleaner to have two different functions in pumphistory.go, the current time-based one, and a separate recordID-based one. If we do that, then the recordID one can include the record with the given ID (as the last one in the list). The oref0 code would have to check for it (discard it if found, and proceed normally, or do the error handling/fallback if it's not there).

alimhassam commented:
I've been running this code for 24h with good success. See associated client code here: https://github.com/openaps/oref0/pull/1090

I'm also comparing the incremental history to the full history at every loop for testing and everything matches so far. I'll try more convoluted scenarios later today.

I'd like to add a different exit code when record is not found. Unless ypu have plan to change exit code globally for pump communication issues vs other problems that could also allow me to use different code path for retry-able comm errors vs errors which need a full refresh.


Keeping the last record in the output would add extra complexity in the client side witch shouldn't really be necessary. (Parse, check assumptions, filter out etc) Much easier to check the exit code rather then inferring an error based on assumptions on the output imo.

On the topic of having different functions, the oref0 code i write currently uses both filters (time+I'd) so it would be useful to still be able to specify both.

alimhassam commented:
It returns all history up to the record id, but it also stops looking at the cutoff time to avoid unnecessary processing.

ecc1 commented:
But doesn't that approach fail in the case where the pump time has been changed?

alimhassam commented:
Yes that's a good point. Maybe It shouldnt use a cutoff as in most cases it's not needed and not necessary. I'll take it off.

alimhassam commented:
> I was thinking it would be cleaner to have two different functions in pumphistory.go, the current time-based one, and a separate recordID-based one. If we do that, then the recordID one can include the record with the given ID (as the last one in the list). 

I'm non sure how to do this without copy pasting the entire History function with a slight change to it, is that acceptable?

ecc1 commented:
I would do it by defining a common internal function that takes a function-valued parameter (a Go closure) to do the actual testing. Then one could pass a variant of the "findSince" function, and the other could pass a "findRecordID" function.

You could go ahead and prototype it with 2 slightly different versions, make sure it still works, and I can do those "Go style" modifications when I apply the patch.

alimhassam commented:
@ecc1 I updated the change here and addressed some of the comments. I didn't do the CRC32, i tried doing it, but i don't think it was worth it.

If you could add the modifications you suggested, and let me if there are any other comments or discussion points, it would be appreciated.

Thanks!
I'm now testing the new version of this code with oref0.

ecc1 commented:
OK, I applied your patches, added some cleanup/factoring, and pushed them to the https://github.com/ecc1/medtronic/tree/pumpHistoryFrom branch. Let me know if it still works OK for you and I'll merge it back into master.

alimhassam commented:
One small note is that when we add the `BGCapture` in openaps.jq oref0 will start uploading all the fakemeter captures to ns. WHich might be an issue for fakemeter users.
Will need to find a way to filter these out.

ecc1 commented:
Doesn't the "select" near the top do this filtering already?

alimhassam commented:
It filters the `BGReceived` records, but not the `BGCapture` records (renamed to `CalBGForPH` in the new version of script).  since these don't have a `MeterID` field..

I guess some users appreciate the upload of real meter bgs to NS. We'll need to either make this a preference somewhere, or better do smart filtering when the BGCapture record has a corresponding filtered out BGReceived record at the same timestamp.

alimhassam commented:
I'm pretty sure fakemeter creates both types of records. Maybe it's different on different pump models?

ecc1 commented:
You're right, I just confirmed that (at least on x22).

They seem to be adjacent records, so one hack could be to look for one adjacent to a BGReceived with meter ID 000000 and remove it.

ecc1 commented:
I agree that it's backward-compatible, but it requires more storage per record in mongo, due to the new "id" field. Could you do a rough estimate of the increase that you see in production? You could fetch the relevant history from the NS API with curl, note its size, and then remove the id field with jq and see how much was saved. If it's more than 10% or so I'd like to get @scottleibrand 's opinion.

alimhassam commented:
From a quick test on 24h worth of my data, it looks like removing the id fields saves about 4% of non-whitespace characters of my treatments data. (365k to 350k). I don't it should be much concerning. 

Note that records used to be larger and contain other fields on the master branch. Also, we should consider the benefits of not having a race condition in our incremental  pump history retrieval vs the additional space cost. cc @scottleibrand let's get scott's opinion.

scottleibrand commented:
That sounds like a pretty tiny size hit to me. Let’s do it.

alimhassam commented:
Great thanks!!
